### PR TITLE
Decompose nested tapes in HybridOps Pt 2

### PR DESCRIPTION
**Pull request by lillian542**
**Context:**
An initial path-of-least-resistance implementation of `decompose` was added in #638, because it was necessary to complete the task at hand in that PR. However, as it wasn't the main focus of that PR, and it was merged with the plan to fix several things if possible. This PR implements the "cleaner", more updated version of the transform. In the initial implementation:

1. There was no testing for the added decomposition of HybridOp tapes (high priority)
2. A `visited` property was temporarily added to the HybridOp class to make things easier to implement, and should be removed (if possible)
3. A lot of the code involved was almost, but not quite, identical to existing code in PL, and the duplication should be reduced (if possible)

**Description of the Change:**
Tests are added, and the code adjusted to fix any issues that came up. The code was refactored to use most of the comparable functionality in PennyLane instead of duplicating it in catalyst, and to remove the need for the `visited` property on `HybridOps`.

**Benefits:**
We made the changes to make this work more cleanly as discussed during the review process for #638. We avoid duplicate code. We have tests that confirm that the decomposition of nested tapes is working as expected.

### CODE REVIEW

**Comment by lillian542** [comment]
[sc-61450]

File: frontend/test/pytest/test_preprocess.py [hunk 1]
```diff
@@ -244,8 +249,387 @@ def circuit(theta: float):
...
+        tape = qml.tape.QuantumScript([NoMatrixMultiControlledX(wires=[0, 1, 2, 3])])
+
+        with pytest.raises(CompileError, match="could not be decomposed, it might be unsupported"):
+            with EvaluationContext(EvaluationMode.QUANTUM_COMPILATION) as ctx:
+                _ = catalyst_decompose(tape, ctx, stopping_condition)
+
+
 class TestTransform:
-    """Test the transforms implemented in Catalyst."""
+    """Teented in Catalyst."""
```
Comment (erick-xanadu) [inline frontend/test/pytest/test_preprocess.py:632]: ?

Comment (lillian542) [reply frontend/test/pytest/test_preprocess.py:632]: Looks like I somehow deleted `st the transforms implem`? 🙈 

Will undo, thanks!

**Comment by erick-xanadu** [comment]
I added ci-build wheels because there is a change that looks for `.so` and one might need to look for `.dylib`.

File: frontend/test/pytest/test_preprocess.py [hunk 2]
```diff
@@ -65,61 +77,49 @@ def preprocess(self, execution_config: ExecutionConfig = DefaultExecutionConfig)
...
 
 class TestPreprocess:
     """Test the preprocessing transforms implemented in Catalyst."""
 
+    @pytest.mark.skipif(
+        not pathlib.Path(
+            get_lib_path("runtime", "RUNTIME_LIB_DIR") + "/libdummy_device.so"
+        ).is_file(),
+        reason="lib_dummydevice.so was not found.",
+    )
```
Comment (erick-xanadu) [inline frontend/test/pytest/test_preprocess.py:122]: I think this PR is great, but I am just a bit worried that these tests might not be run on macOS. Do you know if that is the case?

**Review by grwlf — CHANGES_REQUESTED** [review]
Hi Lillian! I have a minor suggestion for which I am quite sure we need a fix. My request for change is because of it.

I'd like to discuss the following potential problems:
1. The Python generator used in the previous version of the decomposer was probably intended to drop the recursive call limit. If it is for this reason, I would be glad if we keep it. On the other hand, the limit might be big enough to not be bothered. What do you think?
2. Regarding the tests that check the presence or absence of operations in the text representation of mlir. We have a separate lit-tests that use [FileCheck](https://llvm.org/docs/CommandGuide/FileCheck.html) tool for this purpose. Here is the [lit/test_decomposition.py](https://github.com/PennyLaneAI/catalyst/blob/cded645b62e626abb234da689febce0dd9429ced/frontend/test/lit/test_decomposition.py#L182). I would suggest to write all tests with `assert <OP> in mlir` as lit-tests.

File: frontend/test/pytest/test_preprocess.py [hunk 3]
```diff
@@ -244,6 +249,385 @@ def circuit(theta: float):
...
+            qml.RY(y, 0)
+            adjoint(OtherRX(x, 0))
+            return qml.expval(qml.PauliZ(0))
+
+        mlir = qml.qjit(circuit, target="mlir").mlir
+
+        assert "quantum.adjoint" in mlir
+        assert "RX" in mlir
+        assert "RY" in mlir
+        assert "OtherRY" not in mlir
```
[the remark this answers is not in the document]
Comment (lillian542) [reply frontend/test/pytest/test_preprocess.py:357]: Good catch, thanks! This should be "OtherRX"

File: frontend/test/pytest/test_preprocess.py [hunk 4]
```diff
@@ -244,6 +249,385 @@ def circuit(theta: float):
...
+
+            return qml.state()
+
+        # mlir contains expected gate names, and not the unsupported gate names
+        mlir = qml.qjit(circuit, target="mlir").mlir
+        assert "RX" in mlir
+        assert "CNOT" in mlir
+        assert "PhaseShift" in mlir
+        assert "OtherHadamard" not in mlir
+        assert "OtherIsingXX" not in mlir
```
[the remark this answers is not in the document]
Comment (lillian542) [reply frontend/test/pytest/test_preprocess.py:393]: Thanks, I forgot to update the op names to be more specific, will add that now!

File: frontend/test/pytest/test_preprocess.py [hunk 5]
```diff
@@ -65,61 +77,49 @@ def preprocess(self, execution_config: ExecutionConfig = DefaultExecutionConfig)
...
+    """A version of the Hadamard operator that won't be recognized by the QJit device, and will
+    need to be decomposed"""
 
-        system_extension = ".dylib" if platform.system() == "Darwin" else ".so"
-        lib_path = get_lib_path("runtime", "RUNTIME_LIB_DIR") + "/librtd_dummy" + system_extension
-        return "dummy.remote", lib_path
+    @property
+    def name(self):
+        """name"""
+        return "UnknownOp"
```
Comment (lillian542) [inline frontend/test/pytest/test_preprocess.py:87]:
```suggestion
        return "UnknownOp"
```
```suggestion
        return "OtherHadamard"
```

**Comment by lillian542** [comment]
> The Python generator used in the previous version of the decomposer was probably intended to drop the recursive call limit. If it is for this reason, I would be glad if we keep it. On the other hand, the limit might be big enough to not be bothered. What do you think?

For decomposing individual operators, the Python generator used in the previous version is still called - it's just imported directly from PennyLane (`from pennylane.device.preprocess import decompose` and then decompose calls the generator function). In the previous version it was copy-pasted from PennyLane and slightly modified, and with some refactoring we don't need to modify it, so no need for that code to live in catalyst. 

I didn't include a similar thing for recursing into the nested tapes, but I think the limit without the generator _should_ be fine. I'm not certain what the limit is though, or how nested we anticipate circuits being in the next year or so, so let me know if you have another opinion.

> Regarding the tests that check the presence or absence of operations in the text representation of mlir. We have a separate lit-tests that use [FileCheck](https://llvm.org/docs/CommandGuide/FileCheck.html) tool for this purpose. Here is the [lit/test_decomposition.py](https://github.com/PennyLaneAI/catalyst/blob/cded645b62e626abb234da689febce0dd9429ced/frontend/test/lit/test_decomposition.py#L182). I would suggest to write all tests with assert <OP> in mlir as lit-tests.

I'll take a look - that sounds like best practice, but I'm not very familiar with testing outside of `pytest`, so I would need to learn how those work, and I'm not sure if that's the highest priority thing to be doing right now. These tests work, and there are similar ones (in the same test file) that were here before this PR, so it's not introducing a new "problem". 

Maybe better to just make it a separate "cleaning up" ticket for later, and then when we're a bit less pressed for time I can learn how the MLIR tests work and convert the tests added here as well as existing tests to that format.

**Comment by erick-xanadu** [comment]
@lillian542 , I've fixed pylint on main. :)

**Review by erick-xanadu — APPROVED** [review]

File: frontend/test/pytest/test_preprocess.py (hunk 2 above)
Comment (erick-xanadu) [reply frontend/test/pytest/test_preprocess.py:122]: This has been resolved offline. No changes required.

**Review by grwlf — APPROVED** [review]

**Pull request merged**